README
Crypt::OpenSSL::DSA version 0.13
================================

DESCRIPTION

Crypt::OpenSSL::DSA implements the DSA
(Digital Signature Algorithm) signature verification system.

It is a thin XS wrapper to the DSA functions contained in the 
OpenSSL crypto library, located at http://www.openssl.org

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  openssl, available from http://www.openssl.org/

COPYRIGHT AND LICENCE

Copyright (c) 2002 T.J. Mather.  Crypt::OpenSSL::DSA is free software;
you may redistribute it and/or modify it under the same terms as Perl itself. 

Paid support is available directly from the author of this package.
Please see http://www.maxmind.com/app/opensourceservices for more details.


Changes
Revision history for Perl extension Crypt::OpenSSL::DSA.

0.18    2016/11/17
        - OpenSSL 1.1.0 compatibility
        - fix Kwalitee Issues

0.17    2016/10/27
        - Makefile.PL fixing v0.16

0.16    2016/10/27
        - Makefile.PL supports OPENSSL_PREFIX or OPENSSL_LIB+OPENSSL_INCLUDE env variables
        - Makefile.PL tries to find libcrypto via pkg-config
        - Make the files non-executable
        - Doc fixes (grammar)

0.15    2015/02/03
        - #84367 Win32 compatibility patch
        - #80369 fix errors in POD. Mainly just missing =over/=back
        - #80368 Makefile.PL: unneeded -lssl in LIBS

0.14    2012/10/16
        - compiling with -Werror=format-security now works (Jerome Quelin, #48446)
        - Use Digest::SHA instead of Digest::SHA1 (Ansgar Burchardt, #60271)
        - Security fix, CVE-2009-0129: Missing error check in do_verify (Havard Eidnes, #79958)

0.13    2005/10/15
        - made t/90-openssl-compat.t windows compatible (Rob)

0.12    2005/05/22
        - Removed all use of strlen() in DSA.xs so signatures with nulls,
          as commonly generated with sha1, could be signed/verified,
          and added more tests
          (Brad Fitzpatrick <brad@danga.com>)

0.11    2003/01/06
        - Added -DPERL5 to Makefile.PL required for perl-5.8/gcc-3.2

0.10    2002/09/25
        - Fixed Memory leak with signature object.
        - Added set_p, set_q, set_g, set_pub_key, set_priv_key (Olaf Kolkman)
        - Added set_r and set_s methods to Crypt::OpenSSL::DSA::Signature
        - Renamed Signature.pm to Signature.pod

0.04    2002/09/24
        - Fixed bug in verify method
        - Added read_priv_key_str and read_pub_key_str

0.03    2002/02/06
        - Added Crypt::OpenSSL::DSA::Signature interface

0.02    2001/09/24
        - Added more docs
        - seed is now optional argument in generate_parameters

0.01    2001/09/18
        - original version; created by h2xs


lib/Crypt/OpenSSL/DSA.pm
package Crypt::OpenSSL::DSA;

use strict;
use warnings;

require DynaLoader;

use vars qw(@ISA $VERSION);
@ISA = qw(DynaLoader);
$VERSION = '0.17';

bootstrap Crypt::OpenSSL::DSA $VERSION;

sub read_pub_key_str {
  my ($class, $key_str) = @_;
  $class->_load_key(0, $key_str);
}

sub read_priv_key_str {
  my ($class, $key_str) = @_;
  $class->_load_key(1, $key_str);
}

1;
__END__

=head1 NAME

Crypt::OpenSSL::DSA - Digital Signature Algorithm using OpenSSL

=head1 SYNOPSIS

  use Crypt::OpenSSL::DSA;

  # generate keys and write out to PEM files
  my $dsa = Crypt::OpenSSL::DSA->generate_parameters( 512 );
  $dsa->generate_key;
  $dsa->write_pub_key( $filename );
  $dsa->write_priv_key( $filename );

  # using keys from PEM files
  my $dsa_priv = Crypt::OpenSSL::DSA->read_priv_key( $filename );
  my $sig      = $dsa_priv->sign($message);
  my $dsa_pub  = Crypt::OpenSSL::DSA->read_pub_key( $filename );
  my $valid    = $dsa_pub->verify($message, $sig);

  # using keys from PEM strings
  my $dsa_priv = Crypt::OpenSSL::DSA->read_priv_key_str( $key_string );
  my $sig      = $dsa_priv->sign($message);
  my $dsa_pub  = Crypt::OpenSSL::DSA->read_pub_key_str( $key_string );
  my $valid    = $dsa_pub->verify($message, $sig);

=head1 DESCRIPTION

Crypt::OpenSSL::DSA implements the DSA
(Digital Signature Algorithm) signature verification system.

It is a thin XS wrapper to the DSA functions contained in the 
OpenSSL crypto library, located at http://www.openssl.org

=head1 CLASS METHODS

=over 4

=item $dsa = Crypt::OpenSSL::DSA->generate_parameters( $bits, $seed );

Returns a new DSA object and generates the p, q and g
parameters necessary to generate keys.

bits is the length of the prime to be generated; the DSS allows a maximum of 1024 bits.

=item $dsa = Crypt::OpenSSL::DSA->read_params( $filename );

Reads in a parameter PEM file and returns a new DSA object with the p, q and g
parameters necessary to generate keys.

=item $dsa = Crypt::OpenSSL::DSA->read_pub_key( $filename );

Reads in a public key PEM file and returns a new DSA object that can be used
to verify DSA signatures.

=item $dsa = Crypt::OpenSSL::DSA->read_priv_key( $filename );

Reads in a private key PEM file and returns a new DSA object that can be used
to sign messages.

=item $dsa = Crypt::OpenSSL::DSA->read_pub_key_str( $key_string );

Reads in a public key PEM string and returns a new DSA object that can be used
to verify DSA signatures.
The string should include the -----BEGIN...----- and -----END...----- lines.

=item $dsa = Crypt::OpenSSL::DSA->read_priv_key_str( $key_string );

Reads in a private key PEM string and returns a new DSA object that can be used
to sign messages.
The string should include the -----BEGIN...----- and -----END...----- lines.

=back

=head1 OBJECT METHODS

=over 4

=item $dsa->generate_key;

Generates private and public keys, assuming that $dsa is the return
value of generate_parameters.

=item $sig = $dsa->sign( $message );

Signs $message, returning the signature.  Note that $meesage cannot exceed
20 characters in length.

$dsa is the signer's private key.

=item $sig_obj = $dsa->do_sign( $message );

Similar to C<sign>, but returns a L<Crypt::OpenSSL::DSA::Signature> object.

=item $valid = $dsa->verify( $message, $sig );

Verifies that the $sig signature for $message is valid.

$dsa is the signer's public key.

Note: it croaks if the underlying library call returns error (-1).

=item $valid = $dsa->do_verify( $message, $sig_obj );

Similar to C<verify>, but uses a L<Crypt::OpenSSL::DSA::Signature> object.

Note: it croaks if the underlying library call returns error (-1).

=item $dsa->write_params( $filename );

Writes the parameters into a PEM file.

=item $dsa->write_pub_key( $filename );

Writes the public key into a PEM file.

=item $dsa->write_priv_key( $filename );

Writes the private key into a PEM file.

=item $p = $dsa->get_p, $dsa->set_p($p)

Gets/sets the prime number in binary format.

=item $q = $dsa->get_q, $dsa->set_q($q)

Gets/sets the subprime number (q | p-1) in binary format.

=item $g = $dsa->get_g, $dsa->set_g($g)

Gets/sets the generator of subgroup in binary format.

=item $pub_key = $dsa->get_pub_key, $dsa->set_pub_key($pub_key)

Gets/sets the public key (y = g^x) in binary format.

=item $priv_key = $dsa->get_priv_key, $dsa->set_priv_key($priv_key)

Gets/sets the private key in binary format.

=back

=head1 NOTES

L<Crpyt::DSA> is a more mature Perl DSA module, but can be difficult to
install, because of the L<Math::Pari> requirement.

Comments, suggestions, and patches welcome.

=head1 AUTHOR

T.J. Mather, E<lt>tjmather@maxmind.comE<gt>

=head1 COPYRIGHT

Copyright (c) 2002 T.J. Mather.  Crypt::OpenSSL::DSA is free software;
you may redistribute it and/or modify it under the same terms as Perl itself. 

Paid support is available directly from the author of this package.
Please see L<http://www.maxmind.com/app/opensourceservices> for more details.

=head1 SEE ALSO

L<Crypt::OpenSSL::DSA::Signature>

L<Crypt::DSA>, L<Crypt::OpenSSL::RSA>

L<Net::DNS::SEC>

=cut


Makefile.PL
use strict;
use warnings;
use Config;
use ExtUtils::MakeMaker;

my $lib = '-lcrypto';
my $inc = '';
my $nul = $^O eq 'MSWin32' ? 'nul' : '/dev/null';
if (my $dir = $ENV{OPENSSL_PREFIX}) {
  $lib = "-L$dir/lib -lcrypto";
  $inc = "-I$dir/include";
}
elsif ($ENV{OPENSSL_LIB}) {
  $lib = $ENV{OPENSSL_LIB};
  $inc = $ENV{OPENSSL_INCLUDE};
}
elsif (`pkg-config --modversion libcrypto 2>$nul`) {
  $lib = `pkg-config --libs libcrypto 2> $nul`;
  $inc = `pkg-config --cflags libcrypto 2> $nul`;
}
elsif ($^O eq 'MSWin32') {
  $lib = ['-llibeay32'] if $Config{cc} =~ /cl/;
  $lib = ['-leay32']    if $Config{cc} =~ /gcc/;
}

WriteMakefile(
    'NAME'              => 'Crypt::OpenSSL::DSA',
    'VERSION_FROM'      => 'lib/Crypt/OpenSSL/DSA.pm', # finds $VERSION
    'LICENSE'           => 'perl_5',
    'MIN_PERL_VERSION'  => '5.006',
    'PREREQ_PM'         => {}, # e.g., Module::Name => 1.1
    'AUTHOR'            => 'T.J. Mather',
    'LIBS'              => [ $lib ],
    'INC'               => $inc,
    'DEFINE'            => '-DPERL5',   # perl-5.8/gcc-3.2 needs this
    'META_MERGE'        => {
      resources => {
        repository => 'https://github.com/kmx/perl-Crypt-OpenSSL-DSA',
      },
    },
    dist => { TARFLAGS => '--owner=0 --group=0 -cvf' }, 
);
